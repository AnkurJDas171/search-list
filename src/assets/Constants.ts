export const ALERT_MESSAGE = "This user name does not exist! Please specify an existing user name!";
export const ALERT_VISIBLE_TIME = 2000;

export const LIST_TITLES = [
    'Name',
    'Rank',
    'No. of bananas',
    'isSearchedUser?'
];

export const LIST_LENGTH = 10;
