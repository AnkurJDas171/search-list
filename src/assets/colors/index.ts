const colors = {
    // UI components
    BUTTON: 'rgba(111, 202, 186, 1)',
    TEXT_INPUT: 'rgba(243, 243, 243, 1)',
    LIST_TITLE: 'rgba(111, 202, 186, 1)',
    LIST_ODD_ITEM: 'rgba(204, 255, 229, 1)',
    LIST_EVEN_ITEM: 'rgba(236, 252, 245, 1)',
    SEARCH_ITEM: 'rgba(182, 189, 186, 1)',

    // color names
    BLACK: 'rgba(0, 0, 0, 1)',
    BLACK_SEMI_TRANSPARENT: 'rgba(0, 0, 0, 0.2)',
    WHITE: 'rgba(255, 255, 255, 1)',

    // type
    TRANSPARENT: 'rgba(255, 255, 255, 0)'
}

export default colors;
