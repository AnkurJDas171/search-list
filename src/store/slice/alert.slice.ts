import { createSlice } from "@reduxjs/toolkit";

import alertReducer from "../reducer/alertReducer";
import initialAlertState from "../state/Alert";
import { RootState } from "../";

const alertSlice = createSlice({
    name: 'alert',
    initialState: initialAlertState,
    reducers: alertReducer
});

export const { resetAlert, setAlert } = alertSlice.actions

export const selectCount = (state: RootState) => state.alert;

export default alertSlice.reducer;
