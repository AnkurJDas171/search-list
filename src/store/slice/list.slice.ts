import { createSlice } from "@reduxjs/toolkit";

import listReducer from "../reducer/listReducer";
import { RootState } from "../";
import initialListState from "../state/List";

const listSlice = createSlice({
    name: 'list',
    initialState: initialListState,
    reducers: listReducer
});

export const { clearList, getList, getListFromLowestRank, sortList } = listSlice.actions;

export const selectCount = (state: RootState) => state.list;

export default listSlice.reducer;
