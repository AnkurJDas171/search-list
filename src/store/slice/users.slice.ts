import { createSlice } from "@reduxjs/toolkit";

import usersReducer from "../reducer/usersReducer";
import initialUsersState from "../state/Users";
import { RootState } from "../";

const usersSlice = createSlice({
    name: 'users',
    initialState: initialUsersState,
    reducers: usersReducer
})

export const { getUsers } = usersSlice.actions;

export const selectCount = (state: RootState) => state.users;

export default usersSlice.reducer;
