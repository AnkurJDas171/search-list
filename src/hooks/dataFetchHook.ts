import React, { useEffect } from "react";

import { ERROR_ALERT_MESSAGE, dataUrl } from "../assets/Constants";
import { useAppDispatch } from "./reduxHooks";
import { getUsers } from "../store/slice/users.slice";
import { setAlert } from "../store/slice/alert.slice";
import store from "../store";

const fetchUserData = async (storeData: typeof store.dispatch): Promise<void> => {
    try{
        const response = await fetch(dataUrl);
        const data = await response.json();
        storeData(getUsers(data));
    }catch(err){
        console.log("error is: ", err);
        storeData(setAlert(ERROR_ALERT_MESSAGE))
    }
}

const useDataFetch = () => {
    const dispatch = useAppDispatch();

    useEffect(() => {
        fetchUserData(dispatch)
    }, [])
}

export default useDataFetch;
