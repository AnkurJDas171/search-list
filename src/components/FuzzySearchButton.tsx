import React from "react";
import { StyleSheet } from "react-native";

import AppButton from "../UI/AppButton";
import colors from "../assets/colors";
import { ALERT_MESSAGE, buttonTitles } from "../assets/Constants";
import { FuzzySearchButtonType } from "./types";
import { useAppDispatch, useAppSelector } from "../hooks/reduxHooks";
import { RootState } from "../store";
import { UserState } from "../store/state/type";
import { clearList, getFuzzyMatchedList } from "../store/slice/list.slice";
import { isUserPresentInFuzzyMatch } from "../utils";
import { setAlert } from "../store/slice/alert.slice";

const FuzzySearchButton = ({
    searchedName
}: FuzzySearchButtonType): JSX.Element => {
    const users = useAppSelector((state: RootState): UserState => state.users);
    const dispatch = useAppDispatch();

    const handlePress = (): void => {
        if(isUserPresentInFuzzyMatch(searchedName, users)){
            const payload = {
                searchedName,
                users
            }
            dispatch(getFuzzyMatchedList(payload))
        }else{
            dispatch(clearList());
            dispatch(setAlert(ALERT_MESSAGE));
        }
    }

    return (
        <AppButton
            color={colors.BUTTON}
            containerStyle={styles.container}
            handlePress={handlePress}
            isDisable={searchedName.length === 0}
            title={buttonTitles.FUZZY_SEARCH}
            titleColor={colors.WHITE}
        />
    )
}

const styles = StyleSheet.create({
    container: {
        width: "auto"
    }
})

export default FuzzySearchButton
