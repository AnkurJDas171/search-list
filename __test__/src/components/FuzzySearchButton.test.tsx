import { fireEvent, render } from "@testing-library/react-native";
import React from "react";
import FuzzySearchButton from "../../../src/components/FuzzySearchButton";
import * as hooks from "../../../src/hooks/reduxHooks";
import { ALERT_MESSAGE, testID } from "../../../src/assets/Constants";
import * as listActions from "../../../src/store/slice/list.slice";
import * as alertActions from "../../../src/store/slice/alert.slice";
import mockUsersdata from "../../__mock__/data/UsersData";

jest.mock("../../../src/hooks/reduxHooks", () => ({
    useAppDispatch: jest.fn(() => () => { }),
    useAppSelector: jest.fn()
}))

jest.mock("../../../src/store/slice/list.slice", () => ({
    getFuzzyMatchedList: jest.fn(),
    clearList: jest.fn()
}))

jest.mock("../../../src/store/slice/alert.slice", () => ({
    setAlert: jest.fn()
}))

describe("Fuzzy match button", () => {
    it("should render the button", async () => {
        const screen = render(
            <FuzzySearchButton
                searchedName="sample name"
            />
        )

        const button = await screen.findByTestId(testID.APP_BUTTON);

        expect(screen.toJSON()).toMatchSnapshot();
        expect(button).toBeTruthy();
    })

    it("should call getFuzzyMatchedList method on button press when searched name is partially mathed in user data", async () => {
        jest.spyOn(hooks, 'useAppSelector').mockReturnValue(mockUsersdata);
        jest.spyOn(listActions, 'getFuzzyMatchedList');

        const screen = render(
            <FuzzySearchButton
                searchedName="Rica"
            />
        )

        const button = await screen.findByTestId(testID.APP_BUTTON);

        fireEvent.press(button);

        expect(listActions.getFuzzyMatchedList).toHaveBeenCalled();
    })

    it("should call clearList and setAlert methods on button press when searched name does not match with user data", async () => {
        jest.spyOn(hooks, 'useAppSelector').mockReturnValue(mockUsersdata);
        jest.spyOn(listActions, 'clearList');
        jest.spyOn(alertActions, 'setAlert');

        const screen = render(
            <FuzzySearchButton
                searchedName="hello world"
            />
        )

        const button = await screen.findByTestId(testID.APP_BUTTON);

        fireEvent.press(button);

        expect(listActions.clearList).toHaveBeenCalled();
        expect(alertActions.setAlert).toHaveBeenLastCalledWith(ALERT_MESSAGE);
    })
})
