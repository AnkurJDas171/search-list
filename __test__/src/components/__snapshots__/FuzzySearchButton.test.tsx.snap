// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fuzzy match button should render the button 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      {
        "borderRadius": 2,
      },
      [
        {
          "paddingVertical": 5,
          "width": "30%",
        },
        {
          "width": "auto",
        },
      ],
      false,
    ]
  }
  testID="RNE_BUTTON_WRAPPER"
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="app-button"
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(111, 202, 186, 1)",
          "borderColor": "#2089dc",
          "borderRadius": 5,
          "borderWidth": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 8,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 22,
            "fontWeight": "bold",
            "paddingVertical": 1,
            "textAlign": "center",
          }
        }
      >
        Fuzzy Search
      </Text>
    </View>
  </View>
</View>
`;
